crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs

```diff
@@ -75,13 +98,97 @@ impl AlignedChunk {
 ///   UTF-16 len = UTF-8 len - 2
 ///
 /// So UTF-16 offset = UTF-8 offset - count of bytes `>= 0xC0` - count of bytes `>= 0xE0`
+#[expect(dead_code)]
 pub fn build_translations(source_text: &str, translations: &mut Vec<Translation>) {
+    build_translations_and_lines(source_text, translations, None);
+}
+
+/// Build tables of translations from UTF-8 offsets to UTF-16 offsets and line numbers.
+///
+/// Line breaks handled are \r, \n, \r\n (considered as 1 line break), LS, and PS.
+/// LS and PS are Unicode chars, so handled in the cold path for non-ASCII.
+pub fn build_translations_and_lines(
+    source_text: &str,
+    translations: &mut Vec<Translation>,
+    lines: Option<&mut Vec<LineTranslation>>,
+) {
     // Running counter of difference between UTF-8 and UTF-16 offset
     let mut utf16_difference = 0;
 
-    // Closure that processes a slice of bytes
+    // Line tracking
+    let track_lines = lines.is_some();
+    let mut lines = lines;
+    if track_lines {
+        // Add first line starting at offset 0
+        lines.as_mut().unwrap().push(LineTranslation { utf8_offset: 0, utf16_difference });
+    }
+
+    // Closure that processes a slice of bytes for both unicode and line breaks
     let mut process_slice = |slice: &[u8], start_offset: usize| {
-        for (index, &byte) in slice.iter().enumerate() {
+        let mut index = 0;
+        while index < slice.len() {
+            let byte = slice[index];
+
+            // Handle ASCII line breaks first
+            if track_lines {
+                let line_break_len = match byte {
+                    b'\n' => 1,
+                    b'\r' => {
+                        // Check for \r\n
+                        if index + 1 < slice.len() && slice[index + 1] == b'\n' { 2 } else { 1 }
+                    }
+                    _ => 0,
+                };
+
+                if line_break_len > 0 {
+                    let line_end_offset = start_offset + index + line_break_len;
+                    if line_end_offset < source_text.len() {
```

Comment:
I don't think we should have this check. If file ends with a line break, I think need a `LineTranslation` for it, so that converting offset 2 for source `x\n` gives you `line: 2, column: 0`, not `line: 1, column: 2`.

Reply:
thanks, I edited